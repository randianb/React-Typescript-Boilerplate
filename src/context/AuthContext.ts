import React from 'react';

export interface AuthContextInterface {
  checkingSession: boolean;
  token: string | null;
  idToken: string | null;
  expiresAt: number | null;
  isAuthenticated: boolean;
  handleAuthentication: () => void;
  handleLogin: () => void;
  handleLogout: () => void;
}

export const authContextDefaults: AuthContextInterface = {
  checkingSession: false,
  expiresAt: null,
  token: null,
  idToken: null,
  isAuthenticated: false,
  handleAuthentication: () => null,
  handleLogin: () => null,
  handleLogout: () => null
};

export const AuthContext = React.createContext<AuthContextInterface>(
  authContextDefaults
);
