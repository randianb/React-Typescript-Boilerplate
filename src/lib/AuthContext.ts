import React from 'react';

export interface AuthProfile {
  at_hash?: string;
  aud?: string;
  exp?: number;
  iat?: number;
  iss?: string;
  nonce?: string;
  sub?: string;
  locale?: string;
  name?: string;
  picture?: string;
  email?: string;
}

export interface AuthContextInterface {
  checkingSession: boolean;
  token: string | null;
  idToken: string | null;
  expiresAt: number | null;
  isAuthenticated: boolean;
  handleAuthentication: () => void;
  handleLogin: () => void;
  handleLogout: () => void;
}

export const authContextDefaults: AuthContextInterface = {
  checkingSession: false,
  expiresAt: null,
  token: null,
  idToken: null,
  isAuthenticated: false,
  handleAuthentication: () => null,
  handleLogin: () => null,
  handleLogout: () => null
};

export const AuthContext = React.createContext<AuthContextInterface>(
  authContextDefaults
);
