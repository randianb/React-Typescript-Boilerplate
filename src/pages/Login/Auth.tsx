import React from 'react';
import { LockClosedIcon } from '@heroicons/react/solid';
import axios from 'axios';

export interface IFormInputValues {
  username: string;
  password: string;
  isRemember: boolean;
  isOnLogin: boolean;
  token: string;
}

export function getFormValues(islogin: boolean) {
  const storedValues = localStorage.getItem('form');
  if (!storedValues)
    return {
      username: '',
      password: '',
      isRemember: false,
      token: ''
    };
  const val = JSON.parse(storedValues);
  if (islogin && !val.isRemember) {
    return {
      username: '',
      password: '',
      isRemember: false,
      isOnLogin: false,
      token: ''
    };
  }
  return val;
}
type Props = {
  onLogin: () => void,
}
const Auth = ({ onLogin }: Props) => {
  const setupurl = `${process.env.REACT_APP_MES_SERVER_URL}neu-runtime/ef210fff_f87d_478a_b74d_d2355633971f`;
  const [values, setValues] = React.useState<IFormInputValues>(getFormValues(true));

  React.useEffect(() => {
    localStorage.setItem('form', JSON.stringify(values));
  }, [values]);


  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    setValues((previousValues) => ({
      ...previousValues, isOnLogin: true
    }));
    event.preventDefault();
    if (!process.env.REACT_APP_MES_AUTH_URL) { throw new Error('missing REACT_APP_MES_AUTH_URL config'); }
    const authurl = process.env.REACT_APP_MES_AUTH_URL;
    const url = authurl + `oauth/token?grant_type=password&username=${values.username}&password=${values.password}`
    await axios.post(url)
      .then(response => {
        if (response.status == 200) {
          let data = response.data;
          setValues((previousValues) => ({
            ...previousValues, token: data.value
          }));
          setValues((previousValues) => ({
            ...previousValues, isOnLogin: false
          }));
          onLogin();
        }
        else {
          console.log(response)
          setValues((previousValues) => ({
            ...previousValues, isOnLogin: false
          }));
        }
      }).catch(function (error) {
        console.log(error);
        setValues((previousValues) => ({
          ...previousValues, isOnLogin: false
        }));
      });


  }
  function handleChange(
    event: React.ChangeEvent<HTMLInputElement>
  ) {
    event.persist()
    if (event.target.name == 'isRemember') {
      setValues((previousValues) => ({
        ...previousValues,
        [event.target.name]: !values.isRemember
      }));
    }
    else {
      setValues((previousValues) => ({
        ...previousValues,
        [event.target.name]: event.target.value
      }));
    }
  }
  return (
    <>
      {/*
        This example requires updating your template:

        ```
        <html class="h-full bg-gray-50">
        <body class="h-full">
        ```
      */}
      <div className="min-h-full flex items-center justify-center py-12 px-4 sm:px-6 lg:px-8">
        <div className="max-w-md w-full space-y-8">
          <div>
            <img
              className="mx-auto h-12 w-auto"
              src="https://tailwindui.com/img/logos/workflow-mark-indigo-600.svg"
              alt="Workflow"
            />
            <h2 className="mt-6 text-center text-3xl font-extrabold text-gray-900">
              抽检大屏
            </h2>
            <p className="mt-2 text-center text-sm text-gray-600">
              请使用MES用户登录，如果还没有账号，请{' '}
              <a
                href={setupurl}
                className="font-medium text-indigo-600 hover:text-indigo-500"
              >
                在MES注册新的用户
              </a>
            </p>
          </div>
          <form className="mt-8 space-y-6" onSubmit={handleSubmit}>
            <input type="hidden" name="remember" defaultValue="true" />
            <div className="rounded-md shadow-sm -space-y-px">
              <div>
                <label htmlFor="email-address" className="sr-only">
                  用户名
                </label>
                <input
                  id="email-address"
                  onChange={handleChange}
                  value={values.username}
                  name="username"
                  type="text"
                  required
                  className="appearance-none rounded-none relative block w-full px-3 py-2 border border-gray-300 placeholder-gray-500 text-gray-900 rounded-t-md focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 focus:z-10 sm:text-sm"
                  placeholder="用户名"
                />
              </div>
              <div>
                <label htmlFor="password" className="sr-only">
                  密码
                </label>
                <input
                  id="password"
                  name="password"
                  type="password"
                  onChange={handleChange}
                  value={values.password}
                  autoComplete="current-password"
                  required
                  className="appearance-none rounded-none relative block w-full px-3 py-2 border border-gray-300 placeholder-gray-500 text-gray-900 rounded-b-md focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 focus:z-10 sm:text-sm"
                  placeholder="密码"
                />
              </div>
            </div>

            <div className="flex items-center justify-between">
              <div className="flex items-center">
                <input
                  id="remember-me"
                  name="isRemember"
                  onChange={handleChange}
                  defaultChecked={values.isRemember}
                  type="checkbox"
                  className="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
                />
                <label
                  htmlFor="remember-me"
                  className="ml-2 block text-sm text-gray-900"
                >
                  记住我
                </label>
              </div>
            </div>

            <div>
              <button
                type="submit"
                className="group relative w-full flex justify-center py-2 px-4 border border-transparent text-sm font-medium rounded-md text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
              >
                <span className="absolute left-0 inset-y-0 flex items-center pl-3">
                  {values.isOnLogin
                    ? <svg className="animate-spin -ml-1 mr-3 h-5 w-5 text-white" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24">
                      <circle className="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" strokeWidth="4"></circle>
                      <path className="opacity-75" fill="currentColor" d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"></path>
                    </svg>
                    : <LockClosedIcon
                      className="h-5 w-5 text-indigo-500 group-hover:text-indigo-400"
                      aria-hidden="true"
                    />
                  }
                </span>
                登 录
              </button>
            </div>
          </form>
        </div>
      </div>
    </>
  );
}

export default Auth;