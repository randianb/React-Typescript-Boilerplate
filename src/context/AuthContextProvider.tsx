import React from 'react';
import { AuthContext, AuthContextInterface } from './AuthContext';
import { withRouter, RouteComponentProps } from 'react-router';
import auth0 from 'auth0-js';

export interface AuthContextProviderConfig {
  responseType?: string;
  scope?: string;
}

export interface AuthContextProviderProps extends RouteComponentProps {
  config?: AuthContextProviderConfig;
  loginCallback?: () => void;
  logoutCallback?: () => void;
  renewSessionCallback?: () => void;
}

class AuthContextProviderComponent extends React.Component<
  AuthContextProviderProps
> {
  private auth0: auth0.WebAuth;
  private tokenRenewalTimeout: NodeJS.Timeout | null = null;

  constructor(props: AuthContextProviderProps) {
    super(props);

    if (!process.env.REACT_APP_AUTH_DOMAIN) {
      throw new Error('missing REACT_APP_AUTH_DOMAIN config');
    }

    if (!process.env.REACT_APP_AUTH_CLIENT_ID) {
      throw new Error('missing REACT_APP_AUTH_CLIENT_ID config');
    }

    if (!process.env.REACT_APP_AUTH_CALLBACK) {
      throw new Error('missing REACT_APP_AUTH_CALLBACK config');
    }

    this.auth0 = new auth0.WebAuth({
      domain: process.env.REACT_APP_AUTH_DOMAIN,
      clientID: process.env.REACT_APP_AUTH_CLIENT_ID,
      redirectUri: process.env.REACT_APP_AUTH_CALLBACK,
      responseType:
        (props.config && props.config.responseType) || 'token id_token',
      scope: (props.config && props.config.scope) || 'openid'
    });
  }

  public state: AuthContextInterface = {
    token: null,
    idToken: null,
    expiresAt: 0,
    checkingSession: false,
    isAuthenticated: false,
    handleAuthentication: this.handleAuthentication.bind(this),
    login: this.login.bind(this),
    logout: this.logout.bind(this)
  };

  public componentDidMount() {
    if (localStorage.getItem('isLoggedIn') === 'true') {
      this.renewSession();
    }
  }

  public login() {
    this.auth0.authorize();
  }

  public handleAuthentication() {
    const { loginCallback } = this.props;
    this.auth0.parseHash(async (err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        await this.setSession(authResult);
        if (loginCallback) {
          loginCallback();
        }
      }
      this.props.history.push('/');
    });
  }

  public getToken() {
    return this.state.token;
  }

  public getIdToken() {
    return this.state.idToken;
  }

  public async setSession(authResult: auth0.Auth0DecodedHash): Promise<void> {
    localStorage.setItem('isLoggedIn', 'true');

    const expiresAt = authResult.expiresIn
      ? authResult.expiresIn * 1000 + new Date().getTime()
      : 0;

    await new Promise(resolve => {
      this.setState(
        {
          token: authResult.accessToken || null,
          idToken: authResult.idToken || null,
          expiresAt,
          isAuthenticated: true,
          checkingSession: false
        },
        () => {
          this.scheduleRenewal();
          resolve();
        }
      );
    });
  }

  public renewSession() {
    const { renewSessionCallback } = this.props;

    this.setState({
      checkingSession: true
    });

    this.auth0.checkSession({}, async (err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        await this.setSession(authResult);
        if (renewSessionCallback) {
          renewSessionCallback();
        }
      } else if (err) {
        this.logout();
      }
    });
  }

  public logout() {
    const { logoutCallback } = this.props;

    if (this.tokenRenewalTimeout) {
      clearTimeout(this.tokenRenewalTimeout);
    }

    this.setState(
      {
        token: null,
        idToken: null,
        expiresAt: 0,
        isAuthenticated: false,
        checkingSession: false
      },
      () => {
        if (logoutCallback) {
          logoutCallback();
        }
      }
    );

    localStorage.removeItem('isLoggedIn');

    this.auth0.logout({
      returnTo: window.location.origin
    });
  }

  public isAuthenticated() {
    const expiresAt = this.state.expiresAt || 0;
    return new Date().getTime() < expiresAt;
  }

  public scheduleRenewal() {
    const expiresAt = this.state.expiresAt || 0;
    const timeout = expiresAt - Date.now();

    if (timeout > 0) {
      this.tokenRenewalTimeout = setTimeout(() => {
        this.renewSession();
      }, timeout);
    }
  }

  public render() {
    return (
      <AuthContext.Provider value={this.state}>
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

export const AppContextProvider = withRouter(AuthContextProviderComponent);
