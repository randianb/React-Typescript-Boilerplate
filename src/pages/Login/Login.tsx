import React from 'react';
import { AuthContext } from '../../lib/AuthContext';
import Spinner from '../../components/Spinner/Spinner';

class Login extends React.Component {
  public static contextType = AuthContext;

  public componentDidMount() {
    this.context.login();
  }

  public render() {
    return <Spinner className="loginPage" text="Logging in" />;
  }
}

export default Login;
