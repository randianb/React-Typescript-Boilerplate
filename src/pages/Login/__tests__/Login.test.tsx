import React from 'react';
import { Login } from '../Login';
import { render, cleanup } from '@testing-library/react';
import { AuthContext, authContextDefaults } from '../../../context/AuthContext';

afterEach(cleanup);

const auth = authContextDefaults;

describe('Login Page', () => {
  it('Renders Spinner and calls login', () => {
    let called = false;
    auth.handleLogin = () => {
      called = true;
      return null;
    };
    const { container } = render(
      <AuthContext.Provider value={auth}>
        <Login />
      </AuthContext.Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(called).toBeTruthy();
  });
});
