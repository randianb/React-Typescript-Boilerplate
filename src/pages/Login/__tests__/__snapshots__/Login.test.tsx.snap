// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Renders Spinner and calls login 1`] = `
<div
  class="Spinner loginPage"
>
  <div
    class="Spinner-spinner"
  >
    <svg
      viewBox="0 0 64 64"
    >
      <circle
        r="26"
        transform="translate(32,32)"
      />
    </svg>
  </div>
  <div
    class="Spinner-content"
  >
    Logging in
  </div>
</div>
`;
